src/wasm/wasm-s-parser.cpp

```diff
@@ -3345,13 +3352,40 @@ void SExpressionWasmBuilder::parseElem(Element& s, Table* table) {
 
   auto segment = std::make_unique<ElementSegment>(table->name, offset);
   segment->setName(name, hasExplicitName);
-  parseElemFinish(s, segment, i);
+  parseElemFinish(s, segment, i, usesExpressions);
 }
 
 ElementSegment* SExpressionWasmBuilder::parseElemFinish(
-  Element& s, std::unique_ptr<ElementSegment>& segment, Index i) {
-  for (; i < s.size(); i++) {
-    segment->data.push_back(getFunctionName(*s[i]));
+  Element& s,
+  std::unique_ptr<ElementSegment>& segment,
+  Index i,
+  bool usesExpressions) {
+
+  if (usesExpressions) {
+    for (; i < s.size(); i++) {
+      if (!s[i]->isList()) {
+        throw ParseException("expected a ref.* expression.");
+      }
+      auto& inner = *s[i];
+      if (elementStartsWith(inner, ITEM)) {
+        if (inner[1]->isList()) {
+          // (item (ref.func $f))
+          segment->data.push_back(parseExpression(inner[1]));
+        } else {
+          // (item ref.func $f)
+          inner.list().removeAt(0);
+          segment->data.push_back(parseExpression(inner));
+        }
```

Comment:
Are these really both allowed by the spec? That seems unnecessarily complicated :/

Reply:
I totally agree, it doesn't make any sense. The whole `(item ...)` seems unnecessary. But as for this particular form, I've seen it only once in the spec tests here: https://github.com/WebAssembly/spec/blob/master/test/core/elem.wast#L11

I'm not even sure if it's really intended like that or it's just a typo!

Reply:
Looks like it's intentional: https://webassembly.github.io/reference-types/core/text/modules.html#text-elem. Probably better to just implement it like you've done here than to try to change it in the spec or to just not support it at all in Binaryen.

Reply:
Yeah, at least it doesn't come with a high implementation/complexity cost.